rir/src/compiler/opt/loop_invariant.cpp

```diff
@@ -0,0 +1,97 @@
+#include "../analysis/loop_detection.h"
+#include "../pir/pir_impl.h"
+#include "../transform/bb.h"
+#include "../util/cfg.h"
+#include "../util/safe_builtins_list.h"
+#include "../util/visitor.h"
+#include "R/r.h"
+#include "pass_definitions.h"
+
+namespace rir {
+namespace pir {
+
+typedef std::pair<BB*, BB*> BackEdge;
+typedef std::function<bool(Instruction*)> InstrActionPredicate;
+
+auto noEnvironmentTainting = [](Instruction* i) {
+    if (LdFun::Cast(i))
+        return true;
+
+    if (auto call = CallBuiltin::Cast(i))
+        return SafeBuiltinsList::nonObject(call->builtinId);
+
+    if (CallSafeBuiltin::Cast(i))
+        return true;
+
+    if (CallInstruction::CastCall(i))
+        return false;
+
+    return !(i->mayUseReflection());
+};
+
+bool isSafeToHoistLoads(LoopDetection::Loop& loop) {
+    return loop.holdsPropery(noEnvironmentTainting);
```

Comment:
I include `noEnvironmentTainting` inside `isSafeToHoistLoads` but I do not inline it because I think giving it a name makes `isSafeToHoistLoads` easier to read.